libp2p/protocols/relayv2/relayv2.nim

```diff
@@ -0,0 +1,860 @@
+    return
+
+  if msgRcvFromDstOpt.isNone():
+    trace "error reading stop response", msg = msgRcvFromDstOpt
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  let msgRcvFromDst = msgRcvFromDstOpt.get()
+  if msgRcvFromDst.msgType != StopMessageType.Status:
+    trace "unexpected stop response, not a status message",
+          msgType = msgRcvFromDst.msgType
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  if msgRcvFromDst.status.isNone() or
+     msgRcvFromDst.status.get() != RelayV2Status.Ok:
+    trace "relay stop failure", status = msgRcvFromDst.status
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  let resp = HopMessage(msgType: HopMessageType.Status,
+                        status: some(RelayV2Status.Ok))
+  try:
+    await conn.writeLp(encodeHopMessage(resp).buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing relay response", exc=exc.msg
+    return
+
+  trace "relaying connection", src, dst
+
+  proc bridge(conn: Connection, connDst: Connection) {.async.} =
+    const bufferSize = 4096
+    var
+      bufSrcToDst: array[bufferSize, byte]
+      bufDstToSrc: array[bufferSize, byte]
+      futSrc = conn.readOnce(addr bufSrcToDst[0], bufSrcToDst.len)
+      futDst = connDst.readOnce(addr bufDstToSrc[0], bufDstToSrc.len)
+      bytesSendFromSrcToDst = 0
+      bytesSendFromDstToSrc = 0
+      bufRead: int
+
+    while not conn.closed() and not connDst.closed():
+      try:
+        await futSrc or futDst
+        if futSrc.finished():
+          bufRead = await futSrc
+          assert bufRead <= bufferSize
+          bytesSendFromSrcToDst.inc(bufRead)
+          await connDst.write(@bufSrcToDst[0..<bufRead])
+          zeroMem(addr(bufSrcToDst), bufferSize)
+          futSrc = conn.readOnce(addr bufSrcToDst[0], bufSrcToDst.len)
+        if futDst.finished():
+          bufRead = await futDst
+          assert bufRead <= bufferSize
+          bytesSendFromDstToSrc += bufRead
+          await conn.write(bufDstToSrc[0..<bufRead])
+          zeroMem(addr(bufDstToSrc), bufferSize)
+          futDst = connDst.readOnce(addr bufDstToSrc[0], bufDstToSrc.len)
+      except CancelledError as exc:
+        raise exc
+      except CatchableError as exc:
+        if conn.closed() or conn.atEof():
+          trace "relay src closed connection", src
+        if connDst.closed() or connDst.atEof():
+          trace "relay dst closed connection", dst
+        trace "relay error", exc=exc.msg
+        break
+
+    trace "end relaying", bytesSendFromSrcToDst, bytesSendFromDstToSrc
+
+    await futSrc.cancelAndWait()
+    await futDst.cancelAndWait()
+  await bridge(RelayConnection.new(conn, rv2.limit.duration, rv2.limit.data),
+               RelayConnection.new(connDst, rv2.limit.duration, rv2.limit.data))
+
+
+proc heartbeat(rv2: RelayV2) {.async.} =
+  while true:
```

Comment:
I've just merged a `heartbeat` helper in unstable, see here: https://github.com/status-im/nim-libp2p/blob/991549f3913edeac6c34ee0c07bab31365d53c48/libp2p/protocols/pubsub/gossipsub.nim#L570